c10/util/Exception.cpp

```diff
@@ -195,14 +173,64 @@ WarnAlways::~WarnAlways() {
   set_warnAlways(prev_setting);
 }
 
-} // namespace Warning
+} // namespace WarningUtils
 
-void WarningHandler::process(
+Warning::Warning(
     const SourceLocation& source_location,
     const std::string& msg,
-    const bool /*verbatim*/) {
-  LOG_AT_FILE_LINE(WARNING, source_location.file, source_location.line)
-      << "Warning: " << msg << " (function " << source_location.function << ")";
+    const bool verbatim)
+    : source_location_(source_location),
+      msg_(std::move(msg)),
+      verbatim_(verbatim) {}
+
+Warning::Warning(
+    SourceLocation source_location,
+    detail::CompileTimeEmptyString msg,
+    const bool verbatim)
+    : Warning(std::move(source_location), "", verbatim) {}
+
+Warning::Warning(
+    SourceLocation source_location,
+    const char* msg,
+    const bool verbatim)
+    : source_location_(std::move(source_location)),
+      msg_(std::string(msg)),
+      verbatim_(verbatim) {}
+
+Warning::~Warning() {}
+
+const SourceLocation& Warning::source_location() const {
+  return source_location_;
+}
+
+const std::string& Warning::msg() const {
+  return msg_;
+}
+
+bool Warning::verbatim() const {
+  return verbatim_;
+}
+
+void Warning::warn() const {
+  TORCH_INTERNAL_ASSERT(
+      false,
+      "Cannot call `Warning::warn()`, you must use a subclass of `Warning`");
```

Comment:
Should the base warning class be abstract?

Reply:
I don't think so. According to [this page](https://en.cppreference.com/w/cpp/language/abstract_class), "Abstract types cannot be used as parameter types." But I think we need to be able to use `c10::Warning` as a parameter type